### PR TITLE
Add additional metadata to pass context to octant [#2042]

### PR DESCRIPTION
**Pull request by xtreme-vikram-yadav**

### CODE REVIEW

File: plugin/octant.d.ts [hunk 1]
```diff
@@ -113,31 +113,35 @@ export interface DashboardClient {
   /**
    * Get attempts to fetch a resource using the key provided
    * @param key - the key of the object to be fetched
+   * @param metadata - Optional arbitrary contexual data sent to octant
```
Comment (xtreme-vikram-yadav) [inline plugin/octant.d.ts:116]: These are kept optional to keep the existing plugins from becoming incompatible with octant.

**Comment by xtreme-vikram-yadav** [comment]
Related PR: https://github.com/vmware-tanzu/octant/pull/2099

File: plugin/octant.d.ts (hunk 1 above)
Comment (scothis) [reply plugin/octant.d.ts:116]: Do we have any guidance on what the content of the metadata is? An empty bag doesn't help a user to understand how to define the metadata. Perhaps `opts` is a better name

File: plugin/octant.d.ts [hunk 2]
```diff
@@ -113,31 +113,35 @@ export interface DashboardClient {
...
    * @throws will throw an exception if there is an error with the request
    */
-  Get(key: Key): any;
+  Get(key: Key, metadata?: Record<string, any>): any;
```
Comment (chinigo) [inline plugin/octant.d.ts:119]: Octant will serialize these records when it invokes the plugin's `Get` implementation, either via the gRPC interface or via the JS execution environment, correct?

I wonder if we should use a more constrained type for the values than `any`. I worry about encountering edge cases when serializing complicated data types over both of these interfaces (e.g. differences in numeric representations between gRPC and JS, or cycle detection in serialized object graphs, etc).

If we started off by using just a `Record<string, string>`, we could always expand the type in the future. Clawing it back if we encounter edge cases with `any` will be backwards-incompatible.

File: plugin/octant.d.ts (hunk 1 above)
Comment (xtreme-vikram-yadav) [reply plugin/octant.d.ts:116]: Hey @scothis , metadata is essentially key/value store holding arbitrary data which is sent to octant. I don't have a strong preference for this name although opts will still be an empty bag. I called it metadata to match the name it is given when extracted from octant side. If you strongly prefer opts, please confirm and I will rename it as either of them works for me.

Comment (wwitzel3) [reply plugin/octant.d.ts:116]: I think both `opts` and `metadata` are not quite right. This data is only useful to a plugin author that is working with custom version of Octant that has implemented its own ObjectStore. For example a stepping Octant Playback object store for replaying a set of recorded Cluster and Octant events.

I think we want to clearly relay that for most people, you can and should ignore this. If I was using this library and say a field called `metadata` or `opts` I would want to go find out more about how I should use them.

Comment (wwitzel3) [reply plugin/octant.d.ts:116]: I'm leaning towards being verbose, explicit, and to the point of when I read this I immediately go "I don't care"

`@param customStoreContext - key/values to add to the context for object store requests, only useful for custom object stores`

Comment (wwitzel3) [reply plugin/octant.d.ts:116]: Another option here and it does duplicate some information, is to create a `CustomDashboardClient` or something like that .. that is documented but not the default so the TypeScript completion never even shows up for folks using the standard `DashboardClient`

**Review by wwitzel3 — CHANGES_REQUESTED** [review]
I think we need a better name than `metadata`

File: plugin/octant.d.ts (hunk 1 above)
Comment (xtreme-vikram-yadav) [reply plugin/octant.d.ts:116]: Just trying to get a sense of the cascading effects of defining a duplicate of `DashboardClient`. I'm guessing the plugin change inside `octant.d.ts` with default set to `DashboardClient` will be minimal but are we then conditionally picking one type(DashboardClient) or the other(CustomDashboardClient) when rendering the templates when using yeoman-generator. I guess a user will get an option to pick one or the other dashboard client when spinning up with `yeoman-generator`? Or every instance will be defined with something with type `DashboardClient | CustomDashboardClient`. I like the idea of having it separately defined to not affect the community members but It feels like regardless of which way this is implemented, there will be a point when we will need to document and clarify the use of the `customStoreContext` so it feels like just adding an extra parameter should be good enough?

**Review by wwitzel3 — APPROVED** [review]

**Pull request merged**